src/Cassandra.IntegrationTests/Core/TimestampTests.cs

```diff
@@ -0,0 +1,105 @@
+        public void Cluster_Uses_Provided_Timestamp_Generator()
+        {
+            var generator = new TestTimestampGenerator();
+            using (var cluster = Cluster.Builder()
+                                        .WithTimestampGenerator(generator)
+                                        .AddContactPoint(TestCluster.InitialContactPoint)
+                                        .Build())
+            {
+                var session = cluster.Connect();
+                TestHelper.ParallelInvoke(() => session.Execute("SELECT * FROM system.local"), 10);
+                // The driver should use the generator against C* 2.1+
+                Assert.AreEqual(CassandraVersion < new Version(2, 1) ? 0 : 10, generator.GetCounter());
+            }
+        }
+
+        [Test]
+        public void Should_Use_Statement_Timestamp_Precedence_Over_Cluster_Timestamp_Generator()
+        {
+            var generator = new TestTimestampGenerator();
+            using (var cluster = Cluster.Builder()
+                                        .WithTimestampGenerator(generator)
+                                        .AddContactPoint(TestCluster.InitialContactPoint)
+                                        .Build())
+            {
+                var session = cluster.Connect();
+                var stmt = new SimpleStatement("SELECT * FROM system.local");
+                stmt.SetTimestamp(DateTimeOffset.Now);
+                session.Execute(stmt);
+                // The driver should use the generator against C* 2.1+
+                Assert.AreEqual(0, generator.GetCounter());
+            }
+        }
+
+        [Test]
+        public void Should_Use_CQL_Timestamp_Precedence_Over_Cluster_Timestamp_Generator()
+        {
+            var generator = new TestTimestampGenerator();
+            using (var cluster = Cluster.Builder()
+                                        .WithTimestampGenerator(generator)
+                                        .AddContactPoint(TestCluster.InitialContactPoint)
+                                        .Build())
+            {
+                var session = cluster.Connect();
+                session.CreateKeyspace("timestamp_tests",null, false);
+                session.ChangeKeyspace("timestamp_tests");
+                QueryTools.ExecuteSyncNonQuery(session, string.Format(@"
+                CREATE TABLE {0}(
+                id uuid PRIMARY KEY,
+                label text);", "tbl_timestamp_test"));
+                var now = DateTimeOffset.Parse("2017-01-01T00:00:00.000+00:00");
+                var id = Guid.NewGuid();
+                var stmt = new SimpleStatement("INSERT INTO tbl_timestamp_test (id, label) VALUES (?,?) USING TIMESTAMP ?",
+                    id, "test label", now.Ticks);
+                session.Execute(stmt);
+                // The driver should use the generator against C* 2.1+
+                Assert.AreEqual(CassandraVersion < new Version(2, 1) ? 0 : 4, generator.GetCounter());
```

Comment:
If the protocol version is lower than 3, the driver will not generate a new timestamp.
The driver doesn't parse the query and it's unaware if the query contains `USING TIMESTAMP` (provided by the user only).